torchdata/nodes/filter.py

```diff
@@ -0,0 +1,35 @@
+from typing import Any, Callable, Dict, TypeVar, Optional
+from torchdata.nodes import BaseNode
+
+
+T = TypeVar("T")
+
+
+class Filter(BaseNode[T]):
+    """Node that filters items from source node based on predicate function.
+
+    Args:
+        source_node (BaseNode[T]): The source node to filter items from.
+        filter_fn (Callable[[T], bool]): A function that takes an item and returns True if the item
+            should be included, False otherwise.
+    """
+
```

Comment:
I think we can keep track of self._num_yielded in case we need that information. Like we do here: https://github.com/pytorch/data/blob/83d09fff298d5b14c6b42185166cc691ae47f27b/torchdata/nodes/loader.py#L84C24-L84C35